REST API

cheatfate opened the pull request.

cheatfate committed:
Fix sync_manager.nim to return proper status.
Bump REST API dependencies.

cheatfate committed:
Add nim-presto submodule.

cheatfate committed:
Rest API initial implementation.

cheatfate committed:
Fixing issues after bump.

cheatfate committed:
Attempt to find proper readValue.

cheatfate committed:
Change json serialization framework.
Make all calls which are possible.

cheatfate committed:
Fix fork/schedule procedure comment.

cheatfate committed:
Bump nim-presto.

zah committed:
Add a POST /api/nimbus/v1/graffiti API

cheatfate committed:
Simplification of POST/api/nimbus/v1/graffiti and GET/api/nimbus/v1/graffiti call implementation.

arnetheduck commented on beacon_chain/conf.nim:
```diff
@@ -265,6 +265,21 @@ type
+        name: "rest" }: bool
+
+      restPort* {.
+        defaultValue: defaultEth2RpcPort
```
need a fresh default port for rest, unless both json-rpc and rest can be served through the same port?

cheatfate replied:
Now default port is 5052, this is what Lighthouse uses for REST https://lighthouse-book.sigmaprime.io/api-bn.html

arnetheduck commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,743 @@
+logScope: topics = "rest_beaconapi"
+
+type
+  ValidatorTuple = tuple
```
should we call these something else? putting `Tuple` at the end seems a bit unusual

cheatfate replied:
What is unusual for you? https://github.com/status-im/nimbus-eth2/blob/stable/beacon_chain/spec/eth2_apis/callsigs_types.nim

arnetheduck commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,743 @@
+                                                           cstring] =
+  var res: ValidatorFilter
+  for item in filters:
+    if res * item != {}:
```
do we need to be strict here? it seems harmless to repeat a filter - couldn't it just be included in the final set?

cheatfate replied:
Such strict test was made because of specification https://github.com/ethereum/eth2.0-APIs/blob/master/apis/beacon/states/validators.yaml#L28

mratsim replied:
Is it obvious for everyone that `*` for sets is the intersection? Otherwise `intersection` is clearer

cheatfate replied:
Yes it is obvious for everyone that `*` for set types is an intersection.
https://github.com/nim-lang/Nim/blob/devel/lib/system/setops.nim#L52-L55

cheatfate replied:
One more reason for strict test why you think this query string should be correct:
```
http://127.0.0.1:5052/api/eth/v1/beacon/states/head/validators?state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending&state=pending
```

arnetheduck commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,743 @@
+  # https://ethereum.github.io/eth2.0-APIs/#/Beacon/getBlockHeaders
+  router.api(MethodGet, "/api/eth/v1/beacon/headers") do (
+    slot: Option[Slot], parent_root: Option[Eth2Digest]) -> RestApiResponse:
+    return RestApiResponse.jsonError(Http500, "Not implemented yet")
```
let's highight these with a TODO comment so they're easy to grep for

mratsim commented on beacon_chain/conf.nim:
```diff
@@ -265,6 +265,21 @@ type
+      restAddress* {.
+        defaultValue: defaultAdminListenAddress(config)
+        desc: "Listening address of the REST server"
+        name: "rest-address" }: ValidIpAddress
```
Is there a reason to have a different IP on the same node from other services?

cheatfate replied:
There many scenarios could be used, for example VLAN assigned to free customers which are able to query `json-rpc` and/or `metrics`, and paying customers which are able to query `rest`.

mratsim commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,743 @@
+    slot: Slot
+    validators: seq[ValidatorIndex]
+
+  FailureTuple* = tuple
```
I think RestFailure is better otherise it's not obvious that it's for REST especially given that's it's an exported type.

mratsim commented on beacon_chain/rpc/nimbus_rest_api.nim:
```diff
@@ -0,0 +1,272 @@
+
+  PeerStatusTuple* = tuple
+    peerId: PeerID
+    connected: bool
```
Since those types are exported, should we add Rest somewhere in the name to avoid confusing with the other Peer related data structures?

mratsim reviewed (commented):
Regarding testing, does compiling the nimbus_beacon_node tests all codepaths or do we need extra smoke tests?

Does RPC provides everything to create a blocksim using REST?

cheatfate committed:
Add TODO comments for missing implementations.
Change default REST port to use 5052 (Lighthouse).
Add missing checks for maximum amount of validator ids.

cheatfate committed:
Add prefixes "Rest" to public tuples.

cheatfate committed:
Fix `/eth/v1/config/fork_schedule` path.

cheatfate committed:
Fix validator filtering.

cheatfate committed:
Fix ValidatorIndex issues with workaround.

cheatfate committed:
Add partial implementation for /api/eth/v1/beacon/headers

cheatfate committed:
Rebase and fixes.

cheatfate committed:
Fix /api/eth/v1/node/peer_count response.

cheatfate committed:
Fix /api/eth/v1/node/peers response.

cheatfate committed:
Fix /api/eth/v1/node/peers arguments handling.
Fix maximum Epoch handling.

cheatfate committed:
Fix /api/eth/v1/validator/duties/attester/{epoch} request body ValidatorIndex issues.

cheatfate committed:
Add "BETA version" note to the `--rest` server command line parameters descriptions.

zah commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,874 @@
+    state_id: StateIdent) -> RestApiResponse:
+    let bslot =
+      block:
+        if state_id.isErr():
```
Why are the types of the parameters of the API handlers implicitly wrapped in `Result` types? If there was an error during deserialization, perhaps the framework can produce an error response automatically?

cheatfate replied:
Custom REST servers could produce different HTTP error codes/responses. So its impossible to define default/automatic error response handler.

zah commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,874 @@
+                    return RestApiResponse.jsonError(Http400,
+                                              "Incorrect validator index value")
+                  of ValidatorIndexError.UnsupportedValue:
+                    return RestApiResponse.jsonError(Http500,
```
Are these error responses standardized or are they Nimbus-specific?

cheatfate replied:
Specification defines error code 500 as `Internal Server Error`. Because `ValidatorIndex: uint32` is `nimbus` specific, so all values in range `high(uint32) < x < high(uint40)` now returns error 500. Which is now `Internal Server Error`.

zah replied:
My interpretation of the spec is as follows:

1) If the requester sends a malformed request (e.g. a non-numeric validator ID such as "hello"), we respond with status 400 (Bad Request).

2) If the state ID is not leading to a valid state, we respond with 404.

3) If the validator IDs are not leading to valid validators, this validator ID is omitted from the result.

Point 3) is the interesting one here that is effectively in play when `toValidatorIndex` returns `UnsupportedValue` or `TooHighValue`. Such values for a validator index are too high to be realistically part of the beacon state for the foreseeable future and I think they can safely be skipped without requiring any error response (to satisfy the principle of point 3).

It would be good to bring this for discussion in the #apis channels of the Eth2 R&D discord, so the spec can be updated with a clarification regarding the expected behavior.

cheatfate replied:
Sorry, but i dont see any math logic here.

Is `ValidatorIndex` between `2^32` and `2^40` is valid `ValidatorIndex` value? Yes.

It doesn't matter how high this number is. `2^40` is pretty high, while `2^32` is not high enough (this value even do not cover current Earth population), from my calculation with current parameters (which could be changed in future forks) only 102 years needed to pass `int32` value. Why `int32` value? because on `x86` system Nim is able to address only `int32` sequences and arrays. So i think i even made a mistake here and i need to perform check for `2^31`. 

So i really think that we are planting time bomb like Y2K problem.

Right now logic is absolutely correct, `nimbus` do not support values which are bigger then `2^32`, so all the values which are bigger - are unsupported. To avoid problems with such indexes in the future we return `500 Internal server error`. 
Because this is only `nimbus` problem. All the attempts to convert `2^40` value to `ValidatorIndex` will raise an exception. So this procedure was made to sanitize/check arguments before it will be used in queries.

Also if stat following your logic about "too high to be realistically part of the beacon state for the foreseeable future", then error 500 will not be seen by legal users.

cheatfate replied:
I'm not the guy who produced an idea of `ValidatorIndex` to be `uint32` which is also problem for x86 platform because only `int32` indexing are allowed.
cc @arnetheduck @tersec please respond with some thoughts about what we going to do with `ValidatorIndex`.

tersec replied:
Just to put the existing, documented rationales here first:
https://github.com/status-im/nimbus-eth2/blob/a5713c5d84dce951a310729fb5bb57edbc5c1a47/beacon_chain/spec/datatypes/base.nim#L130-L142

and

https://github.com/status-im/nimbus-eth2/blob/a5713c5d84dce951a310729fb5bb57edbc5c1a47/beacon_chain/spec/datatypes/base.nim#L743-L753

cheatfate replied:
So as you can see here we have `ValidatorIndex = uint32` but on x86 platforms we do not support values `2^31 < x < 2^32` too.

cheatfate replied:
And to overflow 2^31 value theoretically we need near 52 years.

tersec replied:
I agree, `uint32` isn't a great choice. By the current logic, `int32` and accepting the 52-year issue is the immediate mitigation on that end. It doesn't address the REST API issue which brought this up, but it at least makes the current approach internally consistent.

Until we move from validator lists being all variations (including `HashList`, `List`, etc) on `seq[Validator]`, changing `ValidatorIndex` just moves the failure point elsewhere and makes it less obvious.

arnetheduck replied:
one thing about `uint32` vs `uint64` is that it does provide significant memory savings - ie we store about 100k of these for every epoch, about 20 copies at least (shuffled validator indices in `EpochRef`)

The way to think about our `uint32` usage is that according to the standard, it's possible that validatorindices up to `2^40` may happen, but we assume it won't any time soon, so we use `uint32` _internally_ until it's plausible that  the chain will go outside this range - our external behavior however should be that we support the larger range _as if_ there were no validators with such indices - as such, @zah has a good point that it's better to simply answer "no such validator", same as if you requested validator with index `2**30` (which currently doesn't exist).

arnetheduck replied:
Another way to put it: validators are part of a "virtual" list with `2**40` elements where some are empty - right now we have 100k of them approximately, so to index into the virtual list, an int32 is enough, and we allocate only that many - the rest are not allocated but we can still reason about them as if they were empty slots in the list.

cheatfate replied:
From my point of view all the values which form external sources should be strictly validated. Such validation in our case is compression `uint64` value to `uint31`. We can't compress to `uint32` even if `ValidatorIndex == uint32` because this index could not be used in x86 platform. So i expect that any value in [2^31, 2^32] passed to `json-rpc` will crash the x86 node.

Its why its important to strictly validate and properly clarify problem for user. We should not hide any information from our consumers. If we are not support values in range `2^31 - 2^40` on x86 and `2^32 - 2^40` on x64 we should respond with proper error message.

tersec replied:
https://github.com/status-im/nimbus-eth2/pull/2485 addresses the x86 part of this.

zah commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,874 @@
+          if (len(keySet) == 0) and (len(indexSet) == 0):
+            true
+          else:
+            let indexFlag =
```
The if statements here seem equivalent to: 

```nim
let indexFlag = len(indexSet) > 0 and ValidatorIndex(index) in indexSet
```

cheatfate replied:
@zah it is not, because validator's id could be `index` or could be `pubkey`. Because we do not have any helpers which will be able to convert `pubkey` to `index` (or i'm missing something). Such simple structure your propose could not be used. Because you should check not only `indexSet` but `keySet` also.

zah replied:
Well, I was commenting specifically about the way the `indexFlag` variable is computed. If we include the `keySet` into the picture, indeed the whole block of code can be simplified to:

```nim
let includeFlag = (len(keySet) == 0) and (len(indexSet) == 0) or
                  (len(indexSet) > 0 and ValidatorIndex(index) in indexSet) or
                  (len(keySet) > 0 and validator.pubkey in keySet)
```

zah commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,874 @@
+              else:
+                false
+            let keyFlag =
+              if len(keySet) > 0:
```
Ditto

cheatfate replied:
https://github.com/status-im/nimbus-eth2/pull/2459#discussion_r609466718

zah commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,874 @@
+                let vres = item.index.toValidatorIndex()
+                if vres.isErr():
+                  case vres.error()
+                  of ValidatorIndexError.TooHighValue:
```
This snippet of code represents out policy regarding how we should handle invalid validator IDs appearing in the request. Right now, it's duplicated in several functions which makes changing the policy harder than necessary. It can be extracted in a shared function or a template to adhere to the DRY principle.

cheatfate replied:
The problem here is that i can't make template which performs `return` statement because it can't be transformed by `{.async.}` macro.

zah commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -0,0 +1,874 @@
+      let res = proposeSignedBlock(node, head, AttachedValidator(), blck)
+      if res == head:
+        node.network.broadcast(getBeaconBlocksTopic(node.forkDigest), blck)
+        return RestApiResponse.jsonError(Http202, "The block failed " &
```
Defining the strings that are copied verbatim from the API spec as proper constants in Nim will reduce anyone's temptation to change them in the code (e.g. to reword them). In will also reduce information duplication in this particular case.

cheatfate replied:
All errors now constants.

zah commented on beacon_chain/rpc/eth2_json_rest_serialization.nim:
```diff
@@ -0,0 +1,341 @@
+  writeValue(writer, value.data)
+
+## Eth1Address
+proc readValue*(reader: var JsonReader[RestJson], value: var Eth1Address) {.
```
Why did you have to define so many overrides for `JsonrReader[RestJson]`. Aren't all these types already serialized in the same way in the "standard" json flavour?

cheatfate replied:
Its because if i use `RestJson` flavour i can't use "standard" flavour. Also i have defined some of the procedures to use more strict version for decoding integers, because original version allows `-`, `_` while decoding integers.

cheatfate committed:
Make error messages to be constants.

cheatfate committed:
Finish error message to constants process.

cheatfate committed:
Simplify validator_id filter's logic.

cheatfate committed:
Proper fix for ValidatorIndex supported values.
